### PR TITLE
graphql and onchain storage support

### LINKED ISSUE #1724: Add graphql and on-chain storage type support
Linked issue: https://github.com/oceanprotocol/market/issues/1711

Support graphql and on-chain data storage types: 

For graphql:
Parameters: 
* type: graphql
* url : server endpoint url
* query: the actual query to be executed
Example:
```
{
  "type": "graphql",
 "url": "https://v4.subgraph.goerli.oceanprotocol.com/subgraphs/name/oceanprotocol/ocean-subgraph",
  "query": """
    	query{
            nfts(orderBy: createdTimestamp,orderDirection:desc){
                 id
                 symbol
                 createdTimestamp
            }
           }
		  """
}
```

For on-chain data:

Parameters:
* type: smartcontract
* chainId : chainId used to query the contract
* address: the smartcontract address
* abi: the function abi (NOT the entire contract abi)
 
Example:
```
{
	"type": "smartcontract",
        "chainId": 80001, // mumbai
	"address": "0x564955E9d25B49afE5Abd66966Ab4Bc9Ad55Fedb",
	"abi": {
			"inputs": [],
			"name": "swapOceanFee",
			"outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
			"stateMutability": "view",
			"type": "function"
		}
}
```

### PR DESCRIPTION
**Pull request by EnzoVezzaro**
- Fixes #1724 .

Changes proposed in this PR:

- Adding GraphQL integration
- Adding OnChain data integration

Current state:
Screenshot 2022-12-08 at 09 53 16

To do: 
- [x] merge IPFS / Arw integration
- [x] change textarea component
-- [x] install codemirror
-- [x] integrate codemirror into component
-- [x] style codemirror for dark/light + small fixes
-- [x] change layout & validation component
-- ~~[ ] find a way to validate graphql query without schema~~
-- [x] create headers component
-- [x] integrate headers on graphql?
-- [x] integrate headers on url?
-- [x] uncomment headers to provider calls? #1554 
- [x] connect both storages to provider
- [x] finish validation
- [x] add to publish form 
- [x] add to edit form 
- [x] fix preview edit form 
- [x] review tooltip text
- [x] fix tab selection when step changes
- [x] cleanup
- [x] Testing
  - [x] unit tests
  - [x] test publish / edit / view
  - [x] test download (free and fixed) / consume / C2D 

______

### Testing User Flow:

Publish (Free and Fixed price):
- publish assets with GraphQL queries (test more than 1) without headers
- publish assets with GraphQL query (test more than 1) with headers (you can use: https://dgraph.io)
- publish assets with SM (smart contract) abi (test more than 1, with different response type)

Asset Details
- all assets should be correctly detailed (check specially the file informacion)

Edit Asset:
- edit storage type of some of your assets. If you published with GraphQL, than change it any of the other and make sure it's all working as expected

Consume
- consume and download all assets

Compute
- compute what published previously

_____

### CODE REVIEW

**Comment by EnzoVezzaro** [comment]
@alexcos20 I'm getting this error when testing 'smartcontract':

Screenshot 2022-12-05 at 10 55 15

Screenshot 2022-12-05 at 10 55 23

Not sure why is malformed, according to the example [here](https://github.com/oceanprotocol/market/issues/1724) should work.

**Comment by EnzoVezzaro** [comment]
hi @bogdanfazakas , can you check please how we're handling the `address`,  `headers`, `query` and `abi` in asset details in edit for example? When we're in preview and in the asset details page, we're not getting values back to be able to show the correct information. I don't see anything related to this in ocean.js, but can you please take a look?

**Comment by jamiehewitt15** [comment]
I'd suggest that this should be merged first into the [release-candidate-09-01-2023](https://github.com/oceanprotocol/market/tree/release-candidate-09-01-2023) branch

**Comment by jamiehewitt15** [comment]
Conflicts need to be resolved

File: content/pages/editMetadata.json [hunk 1]
```diff
@@ -30,44 +30,112 @@
...
+                "placeholder": "query{\n    nfts(\n       orderBy: createdTimestamp,\n       orderDirection:desc\n   ){\n       id\n       symbol\n       createdTimestamp\n   }\n}",
+                "help": "This QUERY will be stored encrypted after publishing.",
+                "type": "codeeditor",
+                "required": true
+              }
+            ]
+          },
+          {
+            "value": "smartcontract", 
+            "title": "Smartcontract",
```
Comment (jamiehewitt15) [inline content/pages/editMetadata.json:114]:
```suggestion
            "title": "Smart Contract",
```

File: src/@utils/order.ts [hunk 2]
```diff
@@ -63,6 +64,13 @@ export async function order(
   const datatoken = new Datatoken(web3)
   const config = getOceanConfig(asset.chainId)
 
+  const filesEncrypted = await getEncryptedFiles(
+    asset.services[0].files,
+    asset.services[0].serviceEndpoint
+  )
+
+  console.log(filesEncrypted)
```
Comment (jamiehewitt15) [inline src/@utils/order.ts:72]: This can be removed 

```suggestion
```

File: src/@utils/provider.ts [hunk 3]
```diff
@@ -38,6 +55,13 @@ export async function initializeProviderForCompute(
   )
 
   try {
+    const filesEncrypted = await getEncryptedFiles(
+      dataset.services[0].files,
+      dataset.services[0].serviceEndpoint
+    )
+
+    console.log(filesEncrypted)
```
Comment (jamiehewitt15) [inline src/@utils/provider.ts:63]: This can be removed 

```suggestion
```

File: src/components/Asset/AssetActions/index.tsx [hunk 4]
```diff
@@ -56,16 +56,28 @@ export default function AssetActions({
...
-              storageType
+              storageType,
+              query,
+              headers,
+              abi,
+              chainId
             )
           : await getFileDidInfo(asset?.id, asset?.services[0]?.id, providerUrl)
 
+        console.log(asset)
```
Comment (jamiehewitt15) [inline src/components/Asset/AssetActions/index.tsx:79]: This can be removed 

```suggestion
```

**Comment by jamiehewitt15** [comment]
This PR also solves: https://github.com/oceanprotocol/market/issues/1844

**Comment by EnzoVezzaro** [comment]
Hi @alexcos20 do we have an example of a graphql query **with headers** on our infrastructure?

**Comment by jamiehewitt15** [comment]
We need to show an error message when the input fails the validation so the user knows how to fix it. For example, the input below should show a message saying "Invalid json input"

image

**Comment by jamiehewitt15** [comment]
I think there should also be some info somewhere explaining to the user that they only need to input part of the ABI, otherwise it looks like they need to paste the whole ABI

**Comment by jamiehewitt15** [comment]
@EnzoVezzaro can you share an example smart contract that you're using to test this? I tried the one that you sent me but it's not working:

image

**Comment by EnzoVezzaro** [comment]
> @EnzoVezzaro can you share an example smart contract that you're using to test this? I tried the one that you sent me but it's not working:
> 
> image

Mmmm ... 🤔 not sure why you're having this. I'm not having this issue with the example smart contract.
Screenshot 2023-01-12 at 08 04 44

are you on the correct chain? you need to be on mumbai for this to work.

**Comment by jamiehewitt15** [comment]
> are you on the correct chain? you need to be on mumbai for this to work.

Yeah, actually I was on the wrong network. Smart contract publishing and consume seems to work well, I think we just need to improve the error messages as I said above

**Comment by EnzoVezzaro** [comment]
> > are you on the correct chain? you need to be on mumbai for this to work.
> 
> Yeah, actually I was on the wrong network. Smart contract publishing and consume seems to work well, I think we just need to improve the error messages as I said above

great 👍🏽 thanks for reviewing

we're only missing[ the headers ](https://github.com/oceanprotocol/market/pull/1794#issuecomment-1378791545)for now, I'm debugging it. But I'm not having too much luck on what's wrong with the asset.

**Comment by AnaLoznianu** [comment]
@EnzoVezzaro we can use this to have queries with headers as well: 
https://pipedream.com/auth/signup?r=/workflows. Through this, you can create a dummy endpoint and print or do whatever you want with the headers. But just to make sure everything is working as expected.

**Comment by EnzoVezzaro** [comment]
> @EnzoVezzaro we can use this to have queries with headers as well: https://pipedream.com/auth/signup?r=/workflows. Through this, you can create a dummy endpoint and print or do whatever you want with the headers. But just to make sure everything is working as expected.

Thanks Ana 👍🏽 great suggestion

**Comment by mihaisc** [comment]
You can't change the method right? from get to post

**Comment by EnzoVezzaro** [comment]
> You can't change the method right? from get to post

Yeah, It's not included. I might have forgotten 🙁 but from the typing I can see it's only available on `url`. 

Screenshot 2023-01-18 at 07 50 31

I'll add it 👍🏽

**Comment by mihaisc** [comment]
i'll check for graphql, it's probably as you said, just for url and i think we need just `get` and `post` , don't think the other ones make any sense

**Comment by mihaisc** [comment]
All seems to be working, just add the http method for url and it's good

**Comment by EnzoVezzaro** [comment]
ready for review @jamiehewitt15 @bogdanfazakas

**Comment by bogdanfazakas** [comment]
> ready for review @jamiehewitt15 @bogdanfazakas

i ve retested this, the only problem i found seems to be with the arweave assets, was able to publish and edit files with this storage type but the consume is not working. The provider calls are for file info and intialize are failing, and the provider errors are the following:
`
2023-01-25 14:51:33 | 2023-01-25 12:51:33 provider-mumbai-58bc97c64b-fng6d ocean_provider.routes.consume[14] DEBUG Validating :{'type': 'arweave', 'userdata': None}
-- | --
2023-01-25 14:51:33 | 2023-01-25 12:51:33 provider-mumbai-58bc97c64b-fng6d ocean_provider.file_types.file_types_factory[14] DEBUG validate_dict passed on {'type': 'arweave', 'userdata': None}
2023-01-25 14:51:33 | 2023-01-25 12:51:33 provider-mumbai-58bc97c64b-fng6d ocean_provider.routes.consume[14] ERROR malformed service files, missing transactionId.
` 
, so not sure if we set something wrong or the problem is not from the market side.

I tested this using the transactions from the realse candidate file, here is an eg of an asset: `did:op:2ada75c05cdb1791bc489bbabeebb9a07d4d59f8f30c3bb15f70c838fb5d7a76`

**Comment by EnzoVezzaro** [comment]
> > ready for review @jamiehewitt15 @bogdanfazakas
> 
> i ve retested this, the only problem i found seems to be with the arweave assets, was able to publish and edit files with this storage type but the consume is not working. The provider calls are for file info and intialize are failing, and the provider errors are the following:
> 
> `2023-01-25 14:51:33	2023-01-25 12:51:33 provider-mumbai-58bc97c64b-fng6d ocean_provider.routes.consume[14] DEBUG Validating :{'type': 'arweave', 'userdata': None}
> 2023-01-25 14:51:33	2023-01-25 12:51:33 provider-mumbai-58bc97c64b-fng6d ocean_provider.file_types.file_types_factory[14] DEBUG validate_dict passed on {'type': 'arweave', 'userdata': None}
> 2023-01-25 14:51:33	2023-01-25 12:51:33 provider-mumbai-58bc97c64b-fng6d ocean_provider.routes.consume[14] ERROR malformed service files, missing transactionId`
> so not sure if we set something wrong or the problem is not from the market side.
> 
> I tested this using the transactions from the realse candidate file, here is an eg of an asset: `did:op:2ada75c05cdb1791bc489bbabeebb9a07d4d59f8f30c3bb15f70c838fb5d7a76`

@bogdanfazakas Thank you for reviewing the other Storage Types, can you retry now?

**Comment by bogdanfazakas** [comment]
> > > ready for review @jamiehewitt15 @bogdanfazakas
> > 
> > 
> > i ve retested this, the only problem i found seems to be with the arweave assets, was able to publish and edit files with this storage type but the consume is not working. The provider calls are for file info and intialize are failing, and the provider errors are the following:
> > `2023-01-25 14:51:33	2023-01-25 12:51:33 provider-mumbai-58bc97c64b-fng6d ocean_provider.routes.consume[14] DEBUG Validating :{'type': 'arweave', 'userdata': None} 2023-01-25 14:51:33	2023-01-25 12:51:33 provider-mumbai-58bc97c64b-fng6d ocean_provider.file_types.file_types_factory[14] DEBUG validate_dict passed on {'type': 'arweave', 'userdata': None} 2023-01-25 14:51:33	2023-01-25 12:51:33 provider-mumbai-58bc97c64b-fng6d ocean_provider.routes.consume[14] ERROR malformed service files, missing transactionId`
> > so not sure if we set something wrong or the problem is not from the market side.
> > I tested this using the transactions from the realse candidate file, here is an eg of an asset: `did:op:2ada75c05cdb1791bc489bbabeebb9a07d4d59f8f30c3bb15f70c838fb5d7a76`
> 
> @bogdanfazakas Thank you for reviewing the other Storage Types, can you retry now?

Retested everything, the consume works well now for all storage types supported, the only issue encountered was that adding an algorithm to the publisher trusted algorithm, with on chain storage type is not working, other types seem to work well. Not sure if market issue or a validation is failing on a backend service, will update this comment if i discover anything new.

**Comment by EnzoVezzaro** [comment]
> Not sure if market issue or a validation is failing on a backend service, will update this comment if i discover anything new.

Mmmm... 🤔 can you share the link to test it too? maybe I can quickly find the issue 👍🏽

**Comment by bogdanfazakas** [comment]
> > Not sure if market issue or a validation is failing on a backend service, will update this comment if i discover anything new.
> 
> Mmmm... 🤔 can you share the link to test it too? maybe I can quickly find the issue 👍🏽

After more debugging it looks to be a checksum difference / error, i was not able to find logs for that when editing compute settings, but i've allowed all algo's on a dataset and tried to initialize for a compute job and that failed with the checksum error.
Not sure if we should keep this issue blocked just for that or handle it in a sepparate issues, we need to compare the cheksum sent by market with the one computed on the  backend side, since it works for all other storage types

**Comment by EnzoVezzaro** [comment]
> Not sure if we should keep this issue blocked just for that or handle it in a separate issues, we need to compare the checksum sent by market with the one computed on the backend side, since it works for all other storage types

@bogdanfazakas I think we should ship this, it's been hanging for a while now. Do we need anything else on the market for this issue (https://github.com/oceanprotocol/provider/issues/595)?

**Comment by bogdanfazakas** [comment]
> > Not sure if we should keep this issue blocked just for that or handle it in a separate issues, we need to compare the checksum sent by market with the one computed on the backend side, since it works for all other storage types
> 
> @bogdanfazakas I think we should ship this, it's been hanging for a while now. Do we need anything else on the market for this issue ([oceanprotocol/provider#595](https://github.com/oceanprotocol/provider/issues/595))?

yep also discussed this with the team today and decided to approve and merge this, and fix the provider issue independently.

**Review by bogdanfazakas — APPROVED** [review]

**Comment by jamiehewitt15** [comment]
I have a small bit of feedback on the visual side - why is the the font styling different for the placeholder text on the example query? It would be better if it has the same styling as all of our other placeholder text

image

File: content/pages/editMetadata.json [hunk 5]
```diff
@@ -64,7 +64,72 @@
...
+                "required": true
+              }
+            ]
+          },
+          {
+            "value": "smartcontract", 
+            "title": "Smart Contract",
+            "label": "Address",
+            "placeholder": "e.g. 0x8149276f275EEFAc110D74AFE8AFECEaeC7d1593",
+            "help": "This ABI will be stored encrypted after publishing. **Please make sure that the endpoint is accessible over the internet and is not protected by a firewall or by credentials.**",
```
[the remark this answers is not in the document]
Comment (kremalicious) [reply content/pages/editMetadata.json:117]: this has been closed without resolving or following the suggestion

File: content/publish/form.json [hunk 6]
```diff
@@ -138,7 +138,72 @@
...
+                "required": false
+              }
+            ]
+          },
+          {
+            "value": "graphql",
+            "title": "Graphql",
+            "label": "URL",
+            "placeholder": "e.g. http://172.15.0.15:8000/subgraphs/name/oceanprotocol/ocean-subgraph",
+            "help": "This URL will be stored encrypted after publishing. **Please make sure that the endpoint is accessible over the internet and is not protected by a firewall or by credentials.**",
```
Comment (kremalicious) [inline content/publish/form.json:160]: same as above, help text does not fully apply to graphql as you are asking for auth details in the next `innerFields`

File: content/publish/form.json [hunk 7]
```diff
@@ -138,7 +138,72 @@
...
+                "required": true
+              }
+            ]
+          },
+          {
+            "value": "smartcontract", 
+            "title": "Smartcontract",
+            "label": "Address",
+            "placeholder": "e.g. 0x8149276f275EEFAc110D74AFE8AFECEaeC7d1593",
+            "help": "This ABI will be stored encrypted after publishing.",
```
Comment (kremalicious) [inline content/publish/form.json:191]: same as on editMetadata.json, why are there multiple `help` keys?

File: package.json [hunk 8]
```diff
@@ -23,13 +23,15 @@
     "storybook:build": "cross-env NODE_ENV=test build-storybook"
   },
   "dependencies": {
+    "@codemirror/language": "^6.3.1",
```
Comment (kremalicious) [inline package.json:26]: all those new packages (do we really need them?) should be under `dependencies` and not scattered between `dependencies` and `devDependencies`

File: src/@utils/ddo.ts [hunk 9]
```diff
@@ -1,4 +1,20 @@
-import { Asset, DDO, Service } from '@oceanprotocol/lib'
+import FormEditMetadata from '@components/Asset/Edit/FormEditMetadata'
```
Comment (kremalicious) [inline src/@utils/ddo.ts:1]: what GitHub says, "`FormEditMetadata` is defined but never used"

File: src/@utils/codemirror.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+export function checkJson(text: string) {
+  try {
+    JSON.parse(text)
+    return true
+  } catch (error) {
+    return false
+  }
+}
+
+export const oceanTheme = (marketTheme: any, field) => {
```
Comment (kremalicious) [inline src/@utils/codemirror.ts:13]: we know `marketTheme` is either `light` or `dark`, which also matches `Theme` typing from `@uiw/codemirror-themes` (sadly they don't export it). So no `any` please.

File: src/@utils/ddo.ts [hunk 11]
```diff
@@ -70,3 +86,90 @@ export function secondsToString(numberOfSeconds: number): string {
     ? `${seconds} second${numberEnding(seconds)}`
     : 'less than a second'
 }
+
+export function normalizeFile(storageType: string, file: any, chainId: number) {
```
Comment (kremalicious) [inline src/@utils/ddo.ts:90]: we should have a typing for that `file` somewhere, right?

File: content/publish/form.json (hunk 7 above)
Comment (EnzoVezzaro) [reply content/publish/form.json:191]: fixed 👍🏽 yeah, they are used by the same component we have in publish.

Screenshot 2023-02-09 at 08 17 38

File: content/publish/form.json (hunk 6 above)
Comment (EnzoVezzaro) [reply content/publish/form.json:160]: good catch, thanks

**Comment by EnzoVezzaro** [comment]
> I have a small bit of feedback on the visual side - why is the the font styling different for the placeholder text on the example query? It would be better if it has the same styling as all of our other placeholder text

I'm not sure since this is a field that is replicating a small code editor. It has the correct font for a code editor.

**Review by mihaisc — APPROVED** [review]

**Pull request merged**